### PR TITLE
Enable testing of -fpatchable-function-entry instrumentation.

### PR DESCRIPTION
**Pull request by yugr**
This PR adds testing configuration for `-fpatchable-function-entry` to `runtests.py`. It mostly fails on GCC due to #1613 but works well with clang (955 tests passed VS 1006 with `-pg`).

### CODE REVIEW

**Review by namhyung — COMMENTED** [review]
Thanks for the PR.  For some reason, it creates a lot of segfaults for `clang` + `-finstrument-functions`.

Also I think you need this to align headers properly

```diff
diff --git a/tests/runtest.py b/tests/runtest.py
index 679d53e20d..be977f0ea8 100755
--- a/tests/runtest.py
+++ b/tests/runtest.py
@@ -799,7 +799,7 @@ result_string = {
     header1 = '%-24s ' % 'Compiler'
     header2 = '%-24s ' % 'Test case'
     header3 = '-' * 24 + ':'
-    empty = '                             '
+    empty = ' ' * 100
 
     for i, compiler in enumerate(compilers):
         if i != 0:
```

**Comment by yugr** [comment]
> Thanks for the PR. For some reason, it creates a lot of segfaults for clang + -finstrument-functions.

Hm, strange, I can't repro on Ubuntu 22. As I mentioned there are a lot of crashes in GCC -fpatchable-function-entry but existing configurations are unchanged.

**Comment by yugr** [comment]
> > Thanks for the PR. For some reason, it creates a lot of segfaults for clang + -finstrument-functions.
> 
> Hm, strange, I can't repro on Ubuntu 22.

Also on Ubuntu 20.

**Comment by yugr** [comment]
@namhyung which distro do you experience those segfaults on? I'll try to install same config locally.

**Comment by namhyung** [comment]
> @namhyung which distro do you experience those segfaults on? I'll try to install same config locally.

It's a custom debian managed by my employer, and I guess it has a number of modifications on the toolchains.  I'll take a look on my side..

**Comment by honggyukim** [comment]
I also see many segfaults as follows.
```
$ ./runtest.py                                                              
Start 280 tests with 6 worker                                               
Compiler                  gcc                                           clang                                       
Test case                 pg             finstrument-fu fpatchable-fun  pg             finstrument-fu fpatchable-fun
------------------------: O0 O1 O2 O3 Os O0 O1 O2 O3 Os O0 O1 O2 O3 Os  O0 O1 O2 O3 Os O0 O1 O2 O3 Os O0 O1 O2 O3 Os
001 basic               : OK OK OK OK OK OK OK OK OK OK NG NG NG NG NG  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
002 argument            : OK OK OK OK OK OK OK OK OK OK NG NG NG NG NG  OK OK OK OK OK SG NZ NZ NZ NZ OK OK OK OK OK
003 thread              : OK OK OK OK OK OK OK OK OK OK NG NG NG NG NG  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
004 filter_F            : OK OK OK OK OK OK OK OK OK OK NZ OK OK OK OK  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
005 filter_N            : OK OK OK OK OK OK OK OK OK OK NG NG NG NG NG  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
006 filter_FN           : OK OK OK OK OK OK OK OK OK OK NZ OK OK OK OK  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
007 library             : OK OK OK OK OK OK OK OK OK OK NG NG NG NG NG  NG NG NG NG NG NG NG NG NG NG NG NG NG NG NG
008 daemon              : OK OK OK OK OK OK OK OK OK OK NG NG NG NG NG  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
009 fork                : OK OK OK OK OK OK OK OK OK OK NG NG NG NG NG  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
010 forkexec            : OK OK OK OK OK OK OK OK OK OK NZ NZ NZ NZ NZ  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
011 vforkexec           : OK OK OK OK OK OK OK OK OK OK NZ NZ NZ NZ NZ  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
012 demangle            : OK OK OK OK OK OK OK OK OK OK NG NG NG NG NG  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
013 signal              : OK OK OK OK OK OK OK OK OK OK NG NG NG NG NG  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
014 ucontext            : OK OK OK OK OK OK OK OK OK OK NG NG NG NG NG  OK NG NG NG NG NZ NZ NZ NZ NZ OK NG NG NG NG
015 longjmp             : OK OK OK OK OK OK OK OK OK OK NG NG NG NG NG  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
016 alloca              : OK OK OK OK OK OK OK OK OK OK NG NG NG NG NG  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
017 no_libcall          : OK OK OK OK OK OK OK OK OK OK NZ NG NG NG NG  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
018 filter_regex        : OK OK OK OK OK OK OK OK OK OK NZ OK OK OK OK  OK OK OK OK OK SG SG SG SG SG OK OK OK OK OK
019 full_depth          : OK OK OK OK OK OK OK OK OK OK NG NG NG NG NG  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
WARN: Segmentation fault: invalid permission (addr: 0x7f7dfd97d130)
WARN:  if this happens only with uftrace, please consider -e/--estimate-return option.

WARN: Backtrace from uftrace v0.13-3-gb8d6 ( x86_64 dwarf python3 luajit tui perf sched dynamic )
WARN: =====================================
WARN: [6] (alloc3[55af76f73415] <= alloc2[55af76f733e9])
WARN: [5] (alloc2[55af76f733c0] <= dynamic_return[7f7dff6b76be])
WARN: [4] (alloc2[55af76f733c5] <= alloc1[55af76f73259])
WARN: [3] (alloc1[55af76f73230] <= dynamic_return[7f7dff6b76be])
WARN: [2] (alloc1[55af76f73235] <= main[55af76f731b3])
WARN: [1] (main[55af76f73170] <= dynamic_return[7f7dff6b76be])
WARN: [0] (main[55af76f73175] <= <7f7dff47ad90>[7f7dff47ad90])

Please report this bug to https://github.com/namhyung/uftrace/issues.

020 filter_depth        : OK OK OK OK OK OK OK OK OK OK NZ OK OK OK OK  OK OK OK OK OK NG NG NG NG NG OK OK OK OK OK
WARN: child terminated by signal: 11: Segmentation fault                                                                                                 
WARN: Segmentation fault: invalid permission (addr: 0x7f4600c50130)                                                                                      
WARN:  if this happens only with uftrace, please consider -e/--estimate-return option.
                                                                                                                                                         
WARN: Backtrace from uftrace v0.13-3-gb8d6 ( x86_64 dwarf python3 luajit tui perf sched dynamic )
WARN: =====================================             
WARN: [6] (alloc3[5617625ec3f5] <= alloc2[5617625ec3cf])
WARN: [5] (alloc2[5617625ec3a0] <= dynamic_return[7f460298a6be])
WARN: [4] (alloc2[5617625ec3a5] <= alloc1[5617625ec23f])
WARN: [3] (alloc1[5617625ec210] <= dynamic_return[7f460298a6be])
WARN: [2] (alloc1[5617625ec215] <= main[5617625ec1a0])
WARN: [1] (main[5617625ec170] <= dynamic_return[7f460298a6be])
WARN: [0] (main[5617625ec175] <= <7f460274dd90>[7f460274dd90])
    ...
```
I also use Ubuntu 22.04.
```
$ cat /etc/os-release | grep PRETTY_NAME
PRETTY_NAME="Ubuntu 22.04.1 LTS"
```

**Comment by honggyukim** [comment]
And many tests fail and they can be fixed by #1615.

**Comment by yugr** [comment]
> I saw many tests fail but they can be fixed by #1615.

Yes, most of GCC -fpatch-function-entry should.

I'm still puzzled about segfaults though. I'll try repro tomorrow on native Linux.

**Comment by yugr** [comment]
Turned out that I needed to install capstone-dev to reproduce crashes.

**Comment by yugr** [comment]
The bug has been fixed, the `-fpatchable-function-entry` tests are mostly green now.

**Comment by namhyung** [comment]
Thanks, it works good now.  But I think you'd better split the changes to
1. fix time range test cases
2. add and use `strip_tracing_flags()`
3. support `-fpatchable-function-entry`

Also you missed the test case 269 to have `-P .` flags for record.

**Comment by yugr** [comment]
> Thanks, it works good now. But I think you'd better split the changes

Right, done. Thanks for spotting the 269 issue.

**Comment by namhyung** [comment]
Thanks for the update.

One minor comment: please add a prefix like `test:` in the commit subject line.  That's a convention to find relevant commits quickly in the future.

**Comment by yugr** [comment]
> Thanks for the update.
> 
> One minor comment: please add a prefix like `test:` in the commit subject line. That's a convention to find relevant commits quickly in the future.

Done.

File: tests/runtest.py [hunk 1]
```diff
@@ -154,6 +154,11 @@ def convert_abs_path(self, build_cmd):
         cmd[src_idx] = abs_src
         return " ".join(cmd)
 
+    def strip_tracing_flags(self, cmd):
+        cmd = cmd.replace('-pg', '').replace('-finstrument-functions', '')
+        cmd = re.sub(r'-fpatchable-function-entry=[0-9]+', '', cmd)
```
Comment (namhyung) [inline tests/runtest.py:160]: Actually the last line belongs to the next change, but it's not a big deal and I don't mind to have it here.

**Review by namhyung — APPROVED** [review]
LGTM

File: tests/runtest.py (hunk 1 above)
Comment (yugr) [reply tests/runtest.py:160]: Good catch, fixed.

**Pull request merged**